Add reject pattern to Admin setting

noellabo opened the pull request:
It was created to refuse a large amount of spam by malicious attackers.
This can be used immediately, but if there is something to improve, please suggest.

noellabo committed:
Add reject pattern to Admin setting

shleeable commented:
If this is about matching strings to reject. Can I request you make this accept a URL with a list of strings?

In principle, we can make a github that people can PR spam strings into collectively.

mjankowski commented:
If you rebase this I can do code review here.

That said, it's been ~10 months with zero feedback from core team ... not sure how viable/desired this is.

mjankowski commented:
@noellabo I know there have been at least SOME other modest anti-spam things created since this ... do you want to rebase and push this forward, or should we close?

noellabo commented:
This pull request was prepared so that Mastodon administrators can cherry-pick it at their own discretion without waiting for the merge.

It is actually being used in the Fedibird fork and is immediately useful, but there has been insufficient discussion about it.

- ReDoS concerns have been raised about using regular expressions
- The option to use external resource reject patterns has been proposed
- Ensure consistency with https://github.com/mastodon/mastodon/pull/32981
- Does it fit with the core team's anti-spam philosophy?

I would like to continue by rebasing.
Please wait a moment.

noellabo commented:
We've rebased and improved code coverage. Please review.

- Currently, the regular expressions are set by the administrator, so they can be inadvertently used to trigger ReDoS, but they can't be used for attacks. Administrators can handle them themselves.
- To keep the patterns we register secret and make it harder for spammers to fight back, we write the reject logic in regular expressions, rather than in source code that needs to be made public.
- We think that incorporating reject patterns from external resources is a separate discussion.
See https://github.com/mastodon/mastodon/issues/29256.
- This pull request filters remote posts.
The anti-spam code at https://github.com/mastodon/mastodon/pull/32981 filters local posts.

This feature was created to provide a measure that does not rely on domain blocking, since many servers tend to domain block servers with a lot of spam during spam outbreaks.

This reject pattern was useful for the Fedibird fork, but after that, spam with attached images started to occur, so I added a rejection function using Blurhash. Furthermore, spam with embedded content in QR codes started to occur, so I added a function to exclude decoded QR codes with a reject pattern. If you are interested, I will make a pull request, so please let me know.

mjankowski commented on app/javascript/material-icons/400-24px/security-fill.svg:
Is this file used?

noellabo replied:
This is unnecessary. I'll delete it.

mjankowski commented on app/lib/activitypub/activity/create.rb:
```diff
@@ -14,7 +14,7 @@ def perform
 
   def create_status
-    return reject_payload! if unsupported_object_type? || non_matching_uri_hosts?(@account.uri, object_uri) || tombstone_exists? || !related_to_local_activity?
+    return reject_payload! if unsupported_object_type? || non_matching_uri_hosts?(@account.uri, object_uri) || tombstone_exists? || !related_to_local_activity? || reject_pattern?
```
The activity/create_spec is comically long, and I see this PR adds spec for populating the added setting value and for the validator - might be useful to also add something to create spec showing the reject payload getting done when we trigger via a reject pattern?

noellabo replied:
Added reject pattern `spam` to activity/create_spec.
- No impact on existing tests
- Added test: Triggered by reject pattern, status not created

noellabo committed:
Delete unuse svg file

noellabo committed:
Add spec activity/create

noellabo committed:
Fix admin_user

ThisIsMissEm commented on app/lib/activitypub/activity/create.rb:
```diff
@@ -452,6 +452,10 @@ def tombstone_exists?
   end
 
+  def reject_pattern?
+    Setting.reject_pattern.present? && @object['content']&.match?(Setting.reject_pattern)
```
I think what you'll see with this is a pivot to spam through the content warning / summary, instead of the content itself, or spam through images attached to the post without any text present.

There's also this that exists: https://github.com/mastodon/mastodon/blob/main/app/lib/antispam.rb

kescherCode commented:
This is the perfect solution to current spam woes. The current phishing spam, the Fediverse Nicole spam, various spam links, are all so easily defeated by catching just one example and putting in into the field.

I've run this branch on my own instance for such a long time now, and given just how much bullshit is spammed from the flagship instance, I really can't fathom how this, or a similar feature, just... aren't added at all. This is basic spam mitigation, and Mastodon just doesn't have it.
